src/services/utilities.ts

```diff
@@ -3663,7 +3663,13 @@ export function createPackageJsonImportFilter(fromFile: SourceFile, preferences:
       ).filter(p => p.parseable);
 
     let usesNodeCoreModules: boolean | undefined;
-    return { allowsImportingAmbientModule, allowsImportingSourceFile, allowsImportingSpecifier };
+    let ambientModuleCache: Map<Symbol, boolean> | undefined;
+    let sourceFileCache: Map<SourceFile, boolean> | undefined;

```

Comment:
I don't think we typically use the symbols or nodes themselves as cache keys; all caches I've dealt with or added use symbol/node IDs instead; not sure if that's intentional or not, though.

Reply:
i think this should be ok now that we are using native maps and life of this cache is well maintained..

Reply:
That does make me wonder if we'd gain anything by switching all of the other caches to work directly...

Reply:
I had a vague memory of saying using the object instead of ids is preferred now. But I don’t know if there’s a real difference.

Reply:
Node IDs are set to `0` unconditionally, and set lazily when we need to use a `Node` in a `Map` or `WeakMap`. I tried to remove some usages out of the compiler and didn't find a ton of wins.

https://github.com/microsoft/TypeScript/pull/46595

The biggest risk here is that you keep a `PackageJsonFilter` around across language service operations - leading to either a memory leak, or corrupt state. A `WeakMap` works, but I'd rather we just be a little careful about how we use these.

So I agree with Sheetal - the cache doesn't escape so we should be okay with these - just make sure the `PackageJsonFilter` object doesn't either.

Reply:
Also - these are `Map`s from things to `boolean` - were you trying to anticipate a refactoring later or something? Otherwise, maybe just use `Set`s.

Reply:
Actually, let me rephrase that - using objects/sparse arrays over `Map`s actually is a win, but not if you rely on iteration order.

Reply:
Yeah, this object only lasts a single request.